lua/null-ls/builtins/formatting/nix_flake_fmt.lua

```diff
@@ -0,0 +1,145 @@
+        --   - `nix eval . --apply '...'`: This doesn't not give me the flake
+        --                                 itself, it gives me the default package.
+        --   - `builtins.getFlake`: Every incantation I've tried requires
+        --                          `--impure`, which has the performance downside described above.
+        --   - `nix flake show --json .`: This works, but it can be quite slow:
+        --                                we end up evaluating all outputs, which can take a while for
+        --                                `nixosConfigurations`.
+        if cp.stderr:find("error: flake .+ does not provide attribute .+ or 'formatter'") then
+            log:warn("this flake does not define a `nix fmt` entrypoint")
+        else
+            log:error(string.format("unable discover 'nix fmt' command. stderr: %s", cp.stderr))
+        end
+
+        return nil
+    end
+
+    if cp.stdout == "" then
+        log:warn("this flake does not define a formatter for your system: %s", nix_current_system)
+        return nil
+    end
+
+    -- stdout has 2 lines of output:
+    --  1. drv path
+    --  2. exe path
+    local drv_path, nix_fmt_path = cp.stdout:match("([^\n]+)\n([^\n]+)\n")
+
+    -- Build the derivation. This ensures that `nix_fmt_path` exists.
+    cp = vim.system({ "nix", "build", "--no-link", drv_path .. "^out" }):wait()
+    if cp.code ~= 0 then
+        log:warn(string.format("unable to build 'nix fmt' entrypoint. stderr: %s", cp.stderr))
+        return nil
+    end
+
+    return nix_fmt_path
+end
+
+return h.make_builtin({
+    name = "nix flake fmt",
+    meta = {
+        url = "https://nix.dev/manual/nix/latest/command-ref/new-cli/nix3-fmt",
+        description = "`nix fmt` - reformat your code in the standard style (this is a generic formatter, not to be confused with nixfmt, a formatter for .nix files)",
+    },
+    method = FORMATTING,
+    filetypes = {},
+    generator_opts = {
+        -- It can take a few moments to find the `nix fmt` entrypoint. The
+        -- underlying command shouldn't change very often for a given
+        -- project, so cache it for the project root.
+        dynamic_command = h.cache.by_bufroot(find_nix_fmt),
+        args = {
+            -- `--walk` is specific to treefmt, and this formatter is supposed
+            -- to be generic.
+            -- Note: this could get converted to the new `TREEFMT_WALK`
+            -- environment variable once
+            -- <https://github.com/numtide/treefmt/commit/bac4a0d102e1142406d3a7d15106e5ba108bfcf8>
+            -- is fixed, which would at least play nicely with other types of
+            -- `nix fmt` entrypoints.
+            --
+            -- However, IMO, the real fix is to change treefmt itself to be
+            -- willing to format files passed explicitly, even if they're
+            -- gitignored:
+            -- https://github.com/numtide/treefmt/issues/435
+            "--walk=filesystem",
```

Comment:
For the record, this is the thing I'd like to figure out before we merge this PR.

Reply:
I just realized I said "gitignored" in this comment. What I really meant was "even if they're not tracked by git".

Reply:
With https://github.com/numtide/treefmt/commit/659aa0fdc20d2deaeaf76d325449ce86d2869036, first released in treefmt v2.1.0, I'm happy to say this is no longer an issue =)